java-checks/src/main/java/org/sonar/java/checks/DateTimeFormatterMismatchCheck.java

```diff
@@ -0,0 +1,145 @@
+    .ofTypes("java.time.format.DateTimeFormatterBuilder")
+    .names("toFormatter")
+    .addParametersMatcher()
+    .build();
+
+  private static final Pattern WEEK_PATTERN = Pattern.compile(".*ww{1,2}.*");
+  private static final Pattern YEAR_OF_ERA_PATTERN = Pattern.compile(".*[uy]+.*");
+
+  private static final String CHANGE_WEEK_FORMAT_MESSAGE = "Change this week format to use the week of week-based year instead.";
+  private static final String CHANGE_YEAR_FORMAT_MESSAGE = "Change this year format to use the week-based year instead.";
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.METHOD_INVOCATION);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    MethodInvocationTree invocation = (MethodInvocationTree) tree;
+    if (OF_PATTERN_MATCHER.matches(invocation)) {
+      visitPattern(invocation);
+    } else if (DATE_TIME_FORMATTER_BUILDER.matches(invocation)) {
+      visitBuildChain(invocation);
+    }
+  }
+
+  private void visitPattern(MethodInvocationTree invocation) {
+    Arguments arguments = invocation.arguments();
+    ExpressionTree argument = arguments.get(0);
+    if (argument.is(Tree.Kind.STRING_LITERAL)) {
+      String pattern = ((LiteralTree) argument).value();
+      if (WEEK_PATTERN.matcher(pattern).matches() && YEAR_OF_ERA_PATTERN.matcher(pattern).matches()) {
+        reportIssue(invocation, CHANGE_YEAR_FORMAT_MESSAGE);
+      }
+    }
+  }
+
+  /**
+   * Walking back through an invocation chain from a call to DateTimeFormatterBuilder.toFormatter looking for calls to appendValue.
+   * If conflicting week and year settings are detected, an issue is reported
+   * @param invocation A call to DateTimeFormatterBuilder.toFormatter
+   */
+  private void visitBuildChain(MethodInvocationTree invocation) {
+    boolean usesWeekBasedYear = false;
+    boolean usesWeekOfWeekBasedYear = false;
+    Tree wanderer = invocation.methodSelect();
+    while (wanderer != null && wanderer.is(Tree.Kind.MEMBER_SELECT)) {
+      ExpressionTree expression = ((MemberSelectExpressionTree) wanderer).expression();
+      if (!expression.is(Tree.Kind.METHOD_INVOCATION)) {
+        break;
+      }
+      MethodInvocationTree mit = (MethodInvocationTree) expression;
+      if (APPEND_VALUE_MATCHER.matches(mit)) {
+        usesWeekBasedYear |= isWeekBasedYearUsed(mit);
+        usesWeekOfWeekBasedYear |= isWeekOfWeekBasedYearUsed(mit);
+      }
+      wanderer = mit.methodSelect();
+    }
+    if (wanderer == null) {
+      return;
+    }
+    ExpressionTree lastExpression = ((MemberSelectExpressionTree) wanderer).expression();
+    if (!lastExpression.is(Tree.Kind.NEW_CLASS)) {
+      return;
+    }
+    if (usesWeekBasedYear && !usesWeekOfWeekBasedYear) {
+      reportIssue(invocation, CHANGE_WEEK_FORMAT_MESSAGE);
+    } else if (!usesWeekBasedYear && usesWeekOfWeekBasedYear) {
+      reportIssue(invocation, CHANGE_YEAR_FORMAT_MESSAGE);
```

Comment:
From the RSPEC, I believe the tree we want to report the issue on is not the `toFormatter` method invocation, but the `appendValue` on which we have the year, with secondary location being the `appendValue` invocation on which the weeks are appended.